### ISSUE TITLE
This package does not adhere fully to JSON:API 🥲 

### ISSUE THREAD

**Issue by Helveg** [ppetzold/nestjs-paginate#753]
# Violations

## [Pagination](https://jsonapi.org/format/#fetching-pagination)

We comply with all the MUSTs, but don't do this SHOULD:

> The page [query parameter family](https://jsonapi.org/format/#query-parameters-families) is reserved for pagination. Servers and clients SHOULD use these parameters for pagination operations.

So we SHOULD change to `page[number]` and `page[limit]` or similar.

## [Sorting](https://jsonapi.org/format/#fetching-sorting)

Here we don't do much of what we MUST do:

* [ ] MUST use the `sort` query parameter
* [ ] MAY support multiple sorts by the comma `,`, applying the sorts in order
* [ ] MUST be ascending, unless the field is prefixed by the minus `-`, which then MUST be descending

## [Filtering](https://jsonapi.org/format/#fetching-filtering)

JSON:API is strategy agnostic, but it does specify the use of the `filter` query parameter family, and that entails that the filter query should be structured like this:

```
?filter[column]=<this piece is already compliant>
```

Since JSON:API does not allow the use of non-family query parameters, we have to change the `search` query parameter to perhaps `filter[@search]` (@ is the only allowed symbol at the start of a member name).

## [Sparse fieldsets](https://jsonapi.org/format/#fetching-sparse-fieldsets)

Currently we use `select`, but this should be changed to `fields[TYPE]` query parameter family, eg `/articles?fields[articles]=title,body`

## [Include relationships](https://jsonapi.org/format/#fetching-includes)

I didn't see from the docs whether we support any syntax for this? But we could support the "Inclusion of Related Resources" as specified.

# Proposal

Short on time, will complete this later next week, but:

* We create backwards-compatible solutions, (and deprecate the old syntax?)
* Or we allow the user to configure whether they want to switch to the new JSON:API syntax

**Comment by Helveg** [comment]
Let me know what you think @ppetzold and I can get started on a branch for this

**Comment by vsamofal** [comment]
@Helveg looks like a good idea 

Do you use some tool that is required to comply with this format?

**Comment by Helveg** [comment]
No, but I like that JSON:API provides a full "HTTP-URL-compliant" spec to query REST API's including nested relationships and field selects! And in general I enjoy adhering to standards because it comes with many benefits :)

**Comment by vsamofal** [comment]
we also have swagger now, it need to be adjusted as well, once you finish I can help with it

**Comment by ppetzold** [comment]
I don't really have a strong opinion on it; as I am currently not having this package in any production use-case. But I see your point, it claims compliance - so it better be compliant lol.

Jsonapi spec seems to be quite active and reasonably popular, so I think I would be okay with taking the leap into full compliance with a breaking change release.

Don't think the headache is worthwhile to maintain backwards compatibility. In case of any security patches we can still patch the current major.

Let's go for a new major with full jsonapi compliance 🚀

**Comment by Helveg** [comment]
Sounds good! Is the PR/release workflow documented somewhere?

**Comment by vsamofal** [comment]
@ppetzold I'm ok as well

**Comment by ppetzold** [comment]
Cool, this repo uses [semantic release](https://semantic-release.gitbook.io/). so every commit (if prefixed) triggers a release.

for larger major releases, we could add a `beta` or `next` branch. but likely overkill for such a small package.

invited both of you to the repo :)

**Comment by Helveg** [comment]
Just one thing, I misinterpreted the query parameter section, and as long as a query parameter contains at least one non a-z character it's a valid "implementation query parameter member name":


>A “query parameter family” is the set of all query parameters whose name starts with a “base name”, followed by zero or more instances of empty square brackets (i.e. []) or square-bracketed legal [member names](https://jsonapi.org/format/#document-member-names). The family is referred to by its base name.
>
>For example, the filter query parameter family includes parameters named: filter, filter[x], filter[], filter[x][], filter[][], filter[x][y], etc. However, filter[_] is not a valid parameter name in the family, because _ is not a valid [member name](https://jsonapi.org/format/#document-member-names).

And:

> Implementation-Specific Query Parameters
>
>Implementations MAY support custom query parameters. However, the names of these query parameters MUST come from a [family](https://jsonapi.org/format/#query-parameters-families) whose base name is a legal [member name](https://jsonapi.org/format/#document-member-names) and also contains at least one non a-z character (i.e., outside U+0061 to U+007A).
>
>It is RECOMMENDED that a capital letter (e.g. camelCasing) be used to satisfy the above requirement.

This means that `filter[@search]` might be needlessly complicated, and we can use a camelCased name instead, like `searchText`, or simply `@search`, but I can't really find a satisfying, intuitive replacement. So if you prefer anything over `filter[@search]` let me know.

**Comment by vsamofal** [comment]
@Helveg 
just a note, it will make sense to release those fixes one by one, having it all in one will be hard to review properly

**Comment by Helveg** [comment]
Ok, I'll PR them separately but targetting a new feature branch, so that they land on `master` as 1 breaking change!

**Comment by orecus** [comment]
Any update or eta on this work and the complaint status? :)

**Comment by Helveg** [comment]
Yes, I'm working on it on the `json-api` branch, and have an open PR https://github.com/ppetzold/nestjs-paginate/pull/883 that addresses some of the points here already.

**Comment by orecus** [comment]
@Helveg @ppetzold Just wanted to check if there is any movement on this and getting this released?

**Comment by Helveg** [comment]
I currently dont have that much open source volunteer time, and am more focused on triaging existing issues. My schedule should open up more in a month or 2
